java/core/src/java/org/apache/orc/impl/InStream.java

```diff
@@ -173,33 +203,208 @@ private static ByteBuffer allocateBuffer(int size, boolean isDirect) {
     }
   }
 
+  /**
+   * Manage the state of the decryption, including the ability to seek.
+   */
+  static class EncryptionState {
+    private final String name;
+    private final EncryptionAlgorithm algorithm;
+    private final Key key;
+    private final byte[] iv;
+    private final Cipher cipher;
+    private ByteBuffer decrypted;
+
+    EncryptionState(String name, StreamOptions options) {
+      this.name = name;
+      algorithm = options.algorithm;
+      key = options.key;
+      iv = options.iv;
+      cipher = algorithm.createCipher();
+    }
+
+    /**
+     * We are seeking to a new range, so update the cipher to change the IV
+     * to match. This code assumes that we only support encryption in CTR mode.
```

Comment:
Comment about how the bottom 8 bytes of IV will be filled (block counter) will be helpful.

Reply:
Ok, I added a comment in CryptoUtils.createIvForStream. In InStream, the IV could be anything they provide although ORC will always have 0 in the bottom 8 bytes.